Make collision with TileMaps useful

HaSa1002 opened the pull request:
Same as #39515 

I just ran into the problem of not beeing able to get the exact colliding tile. Since the described workaround in https://github.com/godotengine/godot/issues/24739#issuecomment-451529763 doesn't work ether and you don't have a chance to get the RID of the quadrant, to get the metadata with the cell position from the Physics2DServer, I gave the TileMap a RID and handle the owners comparable to the CollisionObject2D.

It changes the internal collision handling of the TileMap completly, but otherwise the collision handling of the TileMap is without the changes not usable. You can workaround that, by using use_parent and a StaticBody2D:
```gdscript
func _on_body_shape_entered(body_id, body, body_shape, area_shape):
	if not body is StaticBody2D:
		return
	var b := body as StaticBody2D
	var b_o = b.shape_owner_get_owner(b.shape_find_owner(body_shape))
	if not b_o is TileMap:
		return
	
	var tm : TileMap = b_o
	var cv: Vector2 = Physics2DServer.body_get_shape_metadata(body.get_rid(), body_shape)
```

With the fix you don't need a parent, to be able to use it:

```gdscript
func _on_body_shape_entered(body_id, body, body_shape, area_shape):
	if not body is TileMap:
		return
	var tm := body as TileMap
	var cv := tm.get_cell_from_shape(body_shape)
```

HaSa1002 committed:
Fix collision with TileMaps

This adds the possibility to get position of the collision. It's still
not as straight forward as with normal objects, but now possible.

groud commented:
I don't know whether or not this is useful, but I remember someone doing this by simply throwing a raycast, and using the position of the tip of the raycast to determine which tile it had collided to.

Regarding the code, I cannot tell much, as physics is not my area of expertise. However, if I understand correctly, you are kind of creating a shape (or subshape) for each tile ? If that is the case, we have to make sure so that does not have a significant impact on performance, we cannot afford having tilemaps that make your game lag on mobile.

HaSa1002 commented:
No, I only changed to use one RID for the TileMap instead of one per Quadrant. I don't know if that is impacting performance. I effectivly reimplemented the owner feature of CollisionObject2D into the TileMap and replaced multiple RIDs with one and made it possible to get that one RID. Plus, I added the convienient function, which gathers the TileMap Position that was already stored as metadata in the PhysicsServer, but wasn't accessible, because you can't access the RID of the Quadrants. The problem with the Raycast or any collision is that you get the position of the TileMap not the position of the Tile. With that change you still get the TileMap as collider, but you can get the tile, if you use body_shape_entered using the body_shape parameter, since it is now deterministically stored. Earlier the body_shape id was per Quadrant, which again wasn't accessible.

Chaosus commented:
@groud Does your changes for `TileMap` in 4.0 supersedes this PR?

YuriSizov commented:
@groud Ping!

groud commented:
This is fixed on master using `TileMap.get_coords_for_body_rid(RID p_body_rid)`. This allows retreiving the tile coordinates for a given colliding body.
